File: svc-data/src/main/java/com/beancounter/marketdata/service/MdFactory.java
package com.beancounter.marketdata.service;

import com.beancounter.common.model.Market;
import com.beancounter.marketdata.providers.alpha.AlphaService;
import com.beancounter.marketdata.providers.mock.MockProviderService;
import com.beancounter.marketdata.providers.wtd.WtdService;
import java.util.HashMap;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

/**
 * Return a MarketData provider from a registered collection.
 *
 * @author mikeh
 * @since 2019-03-01
 */
@Service
@Slf4j
public class MdFactory {

  private final Map<String, MarketDataProvider> providers = new HashMap<>();

  MdFactory(MockProviderService mockProviderService,
            AlphaService alphaService,
            WtdService wtdService) {
    providers.put(mockProviderService.getId().toUpperCase(), mockProviderService);
    providers.put(wtdService.getId().toUpperCase(), wtdService);
    providers.put(alphaService.getId().toUpperCase(), alphaService);
  }

  /**
   * Locate a price provider for the requested market.
   *
   * @param market who is pricing this for this market
   * @return the provider that supports the asset
   */
  @Cacheable("providers")
  public MarketDataProvider getMarketDataProvider(Market market) {
    MarketDataProvider provider = resolveProvider(market);
    if (provider == null) {
      return providers.get(MockProviderService.ID);
    }
    return provider;
  }

  public MarketDataProvider getMarketDataProvider(String provider) {
    return providers.get(provider.toUpperCase());
  }

  private MarketDataProvider resolveProvider(Market market) {
    // ToDo: Map Market to Provider
    for (String key : providers.keySet()) {
      if (providers.get(key).isMarketSupported(market)) {
        return providers.get(key);
      }
    }
    log.error("Unable to identify a provider for {}", market);
    return null;
  }


}

File: svc-data/src/main/java/com/beancounter/marketdata/service/MdFactory.kt
package com.beancounter.marketdata.service

import com.beancounter.common.model.Market
import com.beancounter.marketdata.providers.alpha.AlphaService
import com.beancounter.marketdata.providers.mock.MockProviderService
import com.beancounter.marketdata.providers.wtd.WtdService
import org.slf4j.LoggerFactory
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service

/**
 * Return a MarketData provider from a registered collection.
 *
 * @author mikeh
 * @since 2019-03-01
 */
@Service
class MdFactory internal constructor(
    mockProviderService: MockProviderService,
    alphaService: AlphaService,
    wtdService: WtdService
) {
    private val providers: MutableMap<String, MarketDataProvider> = HashMap()

    /**
     * Locate a price provider for the requested market.
     *
     * @param market who is pricing this for this market
     * @return the provider that supports the asset
     */
    @Cacheable("providers")
    fun getMarketDataProvider(market: Market): MarketDataProvider? {
        return resolveProvider(market) ?: return providers[MockProviderService.ID]
    }

    fun getMarketDataProvider(provider: String): MarketDataProvider {
        return providers[provider.toUpperCase()]!!
    }

    private fun resolveProvider(market: Market): MarketDataProvider? {
        // ToDo: Map Market to Provider
        for (key in providers.keys) {
            if (providers[key]!!.isMarketSupported(market)) {
                return providers[key]
            }
        }
        log.error("Unable to identify a provider for {}", market)
        return null
    }

    companion object {
        private val log = LoggerFactory.getLogger(MdFactory::class.java)
    }

    init {
        providers[mockProviderService.getId().toUpperCase()] = mockProviderService
        providers[wtdService.getId().toUpperCase()] = wtdService
        providers[alphaService.getId().toUpperCase()] = alphaService
    }
}
